### ISSUE TITLE
crash on browsing lots of movies and backing out

### ISSUE THREAD

**Issue by cpg** [amahi/android#204]
using the same technique to try and reproduce the crash in #202 with the PR #203 code, i got this crash:
```
FATAL EXCEPTION: main
Process: org.amahi.anywhere, PID: 28899
java.util.concurrent.RejectedExecutionException: Task android.os.AsyncTask$3@7d44d9a rejected from java.util.concurrent.ThreadPoolExecutor@42626cb[Running, pool size = 9, active threads = 9, queued tasks = 128, completed tasks = 735]
    at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2014)
    at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:794)
    at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1340)
    at android.os.AsyncTask.executeOnExecutor(AsyncTask.java:607)
    at org.amahi.anywhere.task.FileMetadataRetrievingTask.execute(FileMetadataRetrievingTask.java:56)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindFileMetadataView(ServerFilesMetadataAdapter.java:127)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindView(ServerFilesMetadataAdapter.java:88)
    at org.amahi.anywhere.adapter.FilesFilterBaseAdapter.getView(FilesFilterBaseAdapter.java:103)
    at android.widget.AbsListView.obtainView(AbsListView.java:2421)
    at android.widget.GridView.makeAndAddView(GridView.java:1440)
    at android.widget.GridView.makeRow(GridView.java:366)
    at android.widget.GridView.fillDown(GridView.java:307)
    at android.widget.GridView.fillGap(GridView.java:267)
    at android.widget.AbsListView.trackMotionScroll(AbsListView.java:5283)
    at android.widget.AbsListView$FlingRunnable.run(AbsListView.java:4831)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:873)
    at android.view.Choreographer.doCallbacks(Choreographer.java:676)
    at android.view.Choreographer.doFrame(Choreographer.java:603)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:859)
    at android.os.Handler.handleCallback(Handler.java:739)
    at android.os.Handler.dispatchMessage(Handler.java:95)
    at android.os.Looper.loop(Looper.java:168)

```

**Comment by cpg** [comment]
@csoni111 said in #203:

> @octacode cleaning the cache won't help! This is happening because the ThreadPoolExecutor has a limit _(afaik its 128 tasks)_ of Asynctask that it can hold at a time. Now since this limit was reached in the current scenario so the `RejectedExecutionException` was thrown.
> This [stackoverflow question](http://stackoverflow.com/questions/34395004/android-app-crashes-when-exceeded-thread-pool-9-and-queued-tasks-128) exactly describes what is happening here. 
> So we will have to think of some mechanism to control the async tasks creation.

This particular movie-share folder does have > 128  movies. And also, I scrolled quite fast, in an effort to see if the issue in #202 was still present in the PR #203 code.

**Comment by csoni111** [comment]
I was just thinking, why do we even need Async Task for fetching Metadata info?
Currently we are using an AsyncTask inside which we are calling retrofit Call's `execute` method. So why not simply use retrofit call's `enqueue` method and remove the async tasks from between. This would solve all the problem but I am not sure if it is the right way!

**Comment by cpg** [comment]
This area is rich in crashes ... same thing ... browse, back, back ... tap, crash ..

```
FATAL EXCEPTION: main
Process: org.amahi.anywhere, PID: 14738
java.lang.RuntimeException: Could not dispatch event: class org.amahi.anywhere.bus.FileMetadataRetrievedEvent to handler [EventHandler public void org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.onFileMetadataRetrieved(org.amahi.anywhere.bus.FileMetadataRetrievedEvent)]: You cannot start a load for a destroyed activity
    at com.squareup.otto.Bus.throwRuntimeException(Bus.java:460)
    at com.squareup.otto.Bus.dispatch(Bus.java:387)
    at com.squareup.otto.Bus.dispatchQueuedEvents(Bus.java:368)
    at com.squareup.otto.Bus.post(Bus.java:337)
    at org.amahi.anywhere.task.FileMetadataRetrievingTask.onResponse(FileMetadataRetrievingTask.java:74)
    at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$1.run(ExecutorCallAdapterFactory.java:68)
    at android.os.Handler.handleCallback(Handler.java:739)
    at android.os.Handler.dispatchMessage(Handler.java:95)
    at android.os.Looper.loop(Looper.java:168)
    at android.app.ActivityThread.main(ActivityThread.java:5885)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687)
 Caused by: java.lang.IllegalArgumentException: You cannot start a load for a destroyed activity
    at com.bumptech.glide.manager.RequestManagerRetriever.assertNotDestroyed(RequestManagerRetriever.java:134)
    at com.bumptech.glide.manager.RequestManagerRetriever.get(RequestManagerRetriever.java:102)
    at com.bumptech.glide.manager.RequestManagerRetriever.get(RequestManagerRetriever.java:87)
    at com.bumptech.glide.Glide.with(Glide.java:629)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindFileMetadataView(ServerFilesMetadataAdapter.java:156)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.bindView(ServerFilesMetadataAdapter.java:145)
    at org.amahi.anywhere.adapter.ServerFilesMetadataAdapter.onFileMetadataRetrieved(ServerFilesMetadataAdapter.java:137)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.squareup.otto.EventHandler.handleEvent(EventHandler.java:89)
    at com.squareup.otto.Bus.dispatch(Bus.java:385)
    at com.squareup.otto.Bus.dispatchQueuedEvents(Bus.java:368) 
    at com.squareup.otto.Bus.post(Bus.java:337) 
    at org.amahi.anywhere.task.FileMetadataRetrievingTask.onResponse(FileMetadataRetrievingTask.java:74) 
    at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$1.run(ExecutorCallAdapterFactory.java:68) 
    at android.os.Handler.handleCallback(Handler.java:739) 
    at android.os.Handler.dispatchMessage(Handler.java:95) 
    at android.os.Looper.loop(Looper.java:168) 
    at android.app.ActivityThread.main(ActivityThread.java:5885) 
    at java.lang.reflect.Method.invoke(Native Method) 
    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797) 
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687) 
```

**Comment by cpg** [comment]
@megabitdragon, this tells me that the browsing may need to have some re-architecting.

**Comment by joydeepsaha05** [comment]
@csoni111 Even Retrofit's enqueue method has a maximum limit to the number of calls in the queue. The limit is 100, so this won't work.

**Comment by csoni111** [comment]
@joydeepsaha05 Where did you find this limit? 
Also, there has to be a limit, you cannot put infinite requests in parallel at the same time. Otherwise this will just hang your device. We only needed a way to control the new requests being made, which retrofit does very smoothly. It handles all of it in the background. So I don't think, there's any issue with that!

**Comment by joydeepsaha05** [comment]
@csoni111 I've noticed the limit when I used Retrofit in one of my other projects. The app would crash on the first run, when all the requests are made for the first time and this would continue for that user, since none of the requests are allowed to complete.

**Comment by csoni111** [comment]
@joydeepsaha05 well I think you might have used an older version of retrofit / okhttp or some other cofiguration, as with the current code it doesn't seem to crash! Also I digged deeper into the code of OkHttp (retrofit uses this to make http calls) and found this in [Dispatcher.java#64](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/Dispatcher.java#L64).

```executorService = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60, TimeUnit.SECONDS, new SynchronousQueue<Runnable>(), Util.threadFactory("OkHttp Dispatcher", false));```

Here they have used Integer.MAX_VALUE for creating the ThreadPoolExecutor. So it should not fail as what you have said!

**Comment by joydeepsaha05** [comment]
@csoni111 Hmm, that might be possible.

But, wouldn't a better solution be to create a single AsyncTask to fetch the metadata one by one (synchronously using Retrofit) and use the onProgressUpdate() to update the UI?

**Comment by csoni111** [comment]
@joydeepsaha05 Say I have 500 movies in a folder _(which is actually the case here)_, if request for loading metadata for each file takes 1 sec. Then it would take 500+ seconds to load the complete data, if you make the requests one by one. This will be really slow. On the other hand if simultaneous requests are fired in separate threads, you can imagine how fast it would be!
Even then you may give it a try and let us know if it works out fine.

Also could you please test the pr #205 and see if you can get the crash log as what @cpg mentioned, because I am unable to figure that one out!

You may get the fake movie list for testing from [here](https://drive.google.com/open?id=0BxF0CJgMqEzkZjR0NDhuQ0J1TTA).

**Comment by joydeepsaha05** [comment]
@csoni111 The earliest I can test things out would be Saturday (Apr 22) since I've got exams this week.

The crash log posted by @cpg is due to the Glide bug which should have been fixed in #202 
Maybe you missed a few files which used the same code?
